### PR TITLE
Guido/23896 psql schema per dataset

### PR DESCRIPTION
**Pull request by gkoller**
❗❗❗ Please see: https://github.com/Amsterdam/schema-tools/pull/237 instead. That PR supercedes this one. ❗❗❗ 

## Introduction

This PR implements https://dev.azure.com/CloudCompetenceCenter/Data%20Diensten/_wiki/wikis/Data-Diensten.wiki/1755/Amsterdam-Schema-Database-wijzigingen with the exception of the copying of data when a new minor version is introduced (last item in the referenced document).

## Features

- Ability to create new datasets in their own DB schema with versioned tables
- Creates updateable views in the public schema for new datasets for backwards compatibility
- Extended the `schema create all` command to be able to work on all datasets. It can now be used, and is recommended, as an alternative for the Django management command `./manage.py import_schemas --create_tables`, that is, for the `--create_tables` part. `import_schemas` on its own is still required for the `dso-api` to "learn" about the datasets it should serve.
- Added `--exclude` option (with the `-x` abbreviation) to `schema create all` to skip certain datasets for when they are broken or concern proxy datasets
- Extend `create sql` cli command with --versioned/no-versioned flag.

## How to use it?

To see this code in action:

### Single dataset

_(This is demonstrated in a short screencast posted in our Slack channel)_

- DROP all `brk` tables from the `public` schema in your `dataservices` database (if any)
- Run `schema create all --db-url postgresql://dataservices:insecure@localhost:5432/dataservices --schema-url ../amsterdam-schema/datasets brk`
- See that:
  - a new DB schema has been created, called `brk`
  - all the tables in that schema have versioned table names without a `brk` prefix, after all, they already live in their own namespace/schema
  - in the `public` schema as many views have been created as tables have been in the `brk` schema. These (updatable) views are basically nothing more than wrapper over the versioned tables in the `brk` schema.

Of course you can use any dataset other than `brk`.

NOTE: the dropping of the existing tables of a given dataset in the `public` schema, for these testing purposes, is necessary. Without it, `schematools` will assume the dataset is a pre-existing one and will **not** create a new DB schema for it. The new functionality in this PR is exclusively for new datasets that do not yet have a presence in the `public` DB schema. 

### Multiple datasets

```console
schema create all --db-url postgresql://dataservices:insecure@localhost:5432/dataservices --exclude corona
```

We're excluding the `corona` dataset as that one is broken; it misses a `crs` property.

## Missing

- No Django code was touched. Hence we should no longer rely on the Django management command to **create** our tables from Jenkins.
- We don't copy over data from one minor version of a table to a newer minor version of that table. This is a potentially expensive process but should be done to realize the vision set out in https://dev.azure.com/CloudCompetenceCenter/Data%20Diensten/_wiki/wikis/Data-Diensten.wiki (see the last section in that document)
- **Please version bump this project after approval en before merge to master.**

## Related PRs:

- https://github.com/Amsterdam/dso-api/pull/452

### CODE REVIEW

File: src/schematools/importer/base.py [hunk 1]
```diff
@@ -463,6 +470,21 @@ def is_versioned_dataset(self) -> bool:
             )
         return is_versioned
 
+    def create_schema(self, db_schema_name: str) -> None:
+        """Create DB Schema.
+
+        Is a no-op if schema already exits.
```
Comment (jjmurre) [inline src/schematools/importer/base.py:534]: exits -> exists

File: src/schematools/importer/base.py [hunk 2]
```diff
@@ -463,6 +470,21 @@ def is_versioned_dataset(self) -> bool:
...
+        """
+        with self.engine.connect() as connection:
+            try:
+                connection.execute(CreateSchema(db_schema_name))
+                self.logger.log_info("Created schema '%s'", db_schema_name)
+            except ProgrammingError as pe:
+                if not isinstance(pe.orig, psycopg2.errors.DuplicateSchema):
+                    # `CreateSchema` does not use the 'IF NOT EXISTS` clause.
+                    # Hence we get an error if the schema already exits.
```
Comment (jjmurre) [inline src/schematools/importer/base.py:543]: same here

File: src/schematools/importer/base.py (hunk 1 above)
Comment (gkoller) [reply src/schematools/importer/base.py:534]: Yeah, I do that often. Unfortunately I lost my access to this repo on Friday December 31st. Hence I can't make any changes anymore.

**Comment by gkoller** [comment]
Can I do anything to facilitate the merge of this PR and that of https://github.com/Amsterdam/dso-api/pull/452?

**Comment by jjmurre** [comment]
> Can I do anything to facilitate the merge of this PR and that of [Amsterdam/dso-api#452](https://github.com/Amsterdam/dso-api/pull/452)?

PR looks really good. I turns out that one of the Airflow operators (table copy) is not compatible with the versioned table approach. But, as long datasets are using the "old" non-versioned approach, that is not a problem and this should not prevent merging this PR.
There is one check I still want to do before merging, that is test some versioned datasets in the DSO API.

**Comment by gkoller** [comment]
@jjmurre I have forked the repo and created a new PR (See https://github.com/Amsterdam/schema-tools/pull/237) so that I was able to rebase my PR on the changes that have happened to `master` after my original PR was created. Saves you to do the rebase and resolve merge conflicts. So it's best probably best to continue in https://github.com/Amsterdam/schema-tools/pull/237 instead of here.

**Pull request closed without merging**